### ISSUE TITLE
Panic caused by contraction

### ISSUE THREAD

**Issue by HdXu** [homotopy-io/homotopy-rs#165]
**Demo**
[panic_4thSep.zip](https://github.com/homotopy-io/homotopy-rs/files/7110322/panic_4thSep.zip)

**Screenshot**
<img width="867" alt="Screenshot 2021-09-04 at 19 25 13" src="https://user-images.githubusercontent.com/35509721/132104574-caea054d-74ee-4b00-ac6a-d32e08bc6e6a.png">

**Description**
When we perform the above contraction, homotopy.io panics.

**Error message**
panicked at 'capacity overflow', library/alloc/src/raw_vec.rs:560:5

Stack:

init/imports.wbg.__wbg_new_59cb74e423758ede@https://beta.homotopy.io/index-7346bddeba3eb5db.js:312:19
console_error_panic_hook::hook::h7c2d1c246b116ea2@https://beta.homotopy.io/index-7346bddeba3eb5db_bg.wasm:wasm-function[1135]:0x15681e
core::ops::function::Fn::call::h9f6eef0081d446f1@https://beta.homotopy.io/index-7346bddeba3eb5db_bg.wasm:wasm-function[2918]:0x1a39aa
std::panicking::rust_panic_with_hook::h2d9079427a2d996a@https://beta.homotopy.io/index-7346bddeba3eb5db_bg.wasm:wasm-function[1856]:0x18e6f7
std::panicking::begin_panic_handler::{{closure}}::h4e49f350d9f4de36@https://beta.homotopy.io/index-7346bddeba3eb5db_bg.wasm:wasm-function[2043]:0x19730d
std::sys_common::backtrace::__rust_end_short_backtrace::h91e711607d91be52@https://beta.homotopy.io/index-7346bddeba3eb5db_bg.wasm:wasm-function[2466]:0x1a134b
rust_begin_unwind@https://beta.homotopy.io/index-7346bddeba3eb5db_bg.wasm:wasm-function[2414]:0x1a0891
core::panicking::panic_fmt::h1fea623755fffed7@https://beta.homotopy.io/index-7346bddeba3eb5db_bg.wasm:wasm-function[2468]:0x1a13af
core::panicking::panic::h38cac710b5048ca0@https://beta.homotopy.io/index-7346bddeba3eb5db_bg.wasm:wasm-function[2373]:0x19fde2
alloc::raw_vec::capacity_overflow::hd612e7efa12c37df@https://beta.homotopy.io/index-7346bddeba3eb5db_bg.wasm:wasm-function[2742]:0x1a316b
alloc::raw_vec::RawVec<T,A>::reserve::do_reserve_and_handle::h0eac977d8a9d3734@https://beta.homotopy.io/index-7346bddeba3eb5db_bg.wasm:wasm-function[1732]:0x18736d
homotopy_graphics::svg::render::GraphicElement::build::hfd2e8dfbbc9345d7@https://beta.homotopy.io/index-7346bddeba3eb5db_bg.wasm:wasm-function[125]:0x115d6
homotopy_web::app::diagram2d::PreparedDiagram::new::hfde73b4b1739ef13@https://beta.homotopy.io/index-7346bddeba3eb5db_bg.wasm:wasm-function[447]:0xe098d
<homotopy_web::app::diagram2d::Diagram2D as yew::html::component::Component>::change::h40274e877a669bcc@https://beta.homotopy.io/index-7346bddeba3eb5db_bg.wasm:wasm-function[218]:0x7fd2f
<yew::html::component::lifecycle::ComponentRunnable<COMP> as yew::scheduler::Runnable>::run::h638f1b1ef13f195a@https://beta.homotopy.io/index-7346bddeba3eb5db_bg.wasm:wasm-function[143]:0x2fb75
std::thread::local::LocalKey<T>::with::h2a7a8165dac3f864@https://beta.homotopy.io/index-7346bddeba3eb5db_bg.wasm:wasm-function[1834]:0x18d26e
yew::scheduler::start::h0fb9306502bde272@https://beta.homotopy.io/index-7346bddeba3eb5db_bg.wasm:wasm-function[2921]:0x1a39c5
<homotopy_web::app::diagram2d::Diagram2D as yew::html::component::Component>::view::{{closure}}::hec6f9ad90d764810@https://beta.homotopy.io/index-7346bddeba3eb5db_bg.wasm:wasm-function[1950]:0x19343d
<yew::html::listener::events::ontransitionend::Wrapper as yew::virtual_dom::Listener>::attach::{{closure}}::he02ebfe89f35ed0c@https://beta.homotopy.io/index-7346bddeba3eb5db_bg.wasm:wasm-function[1928]:0x1922ad
<dyn core::ops::function::FnMut<(&A,)>+Output = R as wasm_bindgen::closure::WasmClosure>::describe::invoke::h1b4525dfe34ec7f4@https://beta.homotopy.io/index-7346bddeba3eb5db_bg.wasm:wasm-function[2412]:0x1a080c
__wbg_adapter_30@https://beta.homotopy.io/index-7346bddeba3eb5db.js:230:14
real@https://beta.homotopy.io/index-7346bddeba3eb5db.js:187:20

index-7346bddeba3eb5db.js:324:21

Uncaught RuntimeError: unreachable executed
    __wbg_adapter_30 https://beta.homotopy.io/index-7346bddeba3eb5db.js:230
    real https://beta.homotopy.io/index-7346bddeba3eb5db.js:187
index-7346bddeba3eb5db_bg.wasm:1719333:1

**Comment by jamievicary** [comment]
Great thanks for this Hao!

Would it be easy for you to make a more minimal example where the diagram
itself is smaller?

**Comment by HdXu** [comment]
**Simplified version**
[panic_5thSep.zip](https://github.com/homotopy-io/homotopy-rs/files/7111826/panic_5thSep.zip)

**Error message**
panicked at 'capacity overflow', library/alloc/src/raw_vec.rs:560:5

Stack:

init/imports.wbg.__wbg_new_59cb74e423758ede@https://beta.homotopy.io/index-7346bddeba3eb5db.js:312:19
console_error_panic_hook::hook::h7c2d1c246b116ea2@https://beta.homotopy.io/index-7346bddeba3eb5db_bg.wasm:wasm-function[1135]:0x15681e
core::ops::function::Fn::call::h9f6eef0081d446f1@https://beta.homotopy.io/index-7346bddeba3eb5db_bg.wasm:wasm-function[2918]:0x1a39aa
std::panicking::rust_panic_with_hook::h2d9079427a2d996a@https://beta.homotopy.io/index-7346bddeba3eb5db_bg.wasm:wasm-function[1856]:0x18e6f7
std::panicking::begin_panic_handler::{{closure}}::h4e49f350d9f4de36@https://beta.homotopy.io/index-7346bddeba3eb5db_bg.wasm:wasm-function[2043]:0x19730d
std::sys_common::backtrace::__rust_end_short_backtrace::h91e711607d91be52@https://beta.homotopy.io/index-7346bddeba3eb5db_bg.wasm:wasm-function[2466]:0x1a134b
rust_begin_unwind@https://beta.homotopy.io/index-7346bddeba3eb5db_bg.wasm:wasm-function[2414]:0x1a0…
index-7346bddeba3eb5db.js:324:21

Uncaught RuntimeError: unreachable executed
index-7346bddeba3eb5db_bg.wasm:1719333:1

**Comment by calintat** [comment]
This seems to be an error in type checking. I will see if I can narrow it down.

**Comment by jamievicary** [comment]
Great to see this fast progress on Hao's issue, thanks Calin.

**Comment by calintat** [comment]
Actually I was wrong, this is an issue in rendering. Contraction works fine but rendering the contracted diagram fails.

The problem is here:
https://github.com/homotopy-io/homotopy-rs/blob/f75d62c725ca66e866127df4da26235514c6248b/homotopy-graphics/src/svg/render.rs#L268-L286
The while loop gets stuck in a cycle. It's a simple fix, but now I'm getting a weird looking result:

<img width="470" alt="Screenshot 2021-09-05 at 21 03 35" src="https://user-images.githubusercontent.com/16837293/132136913-fc16627b-7d91-4941-8b71-2cfa213a7bc5.png">

**Comment by calintat** [comment]
I pushed the fix to unblock Hao and I'm going to make a new issue to investigate the rendering issue with the contracted diagram.

**Comment by calintat** [comment]
The new issue is #166.
